File: src/main/java/arcana/events/WorldEvents.java
package arcana.events;

import arcana.worldgen.ModFeatures;
import arcana.worldgen.Tower;
import net.minecraft.util.RegistryKey;
import net.minecraft.util.registry.Registry;
import net.minecraft.world.World;
import net.minecraft.world.biome.Biome;
import net.minecraft.world.gen.FlatChunkGenerator;
import net.minecraft.world.gen.settings.StructureSeparationSettings;
import net.minecraft.world.server.ServerWorld;
import net.minecraftforge.common.BiomeDictionary;
import net.minecraftforge.event.world.BiomeLoadingEvent;
import net.minecraftforge.event.world.WorldEvent;
import net.minecraftforge.eventbus.api.SubscribeEvent;
import net.minecraftforge.fml.common.Mod;


@Mod.EventBusSubscriber(bus = Mod.EventBusSubscriber.Bus.FORGE)
public class WorldEvents {
    @SubscribeEvent
    static void loadBiome(BiomeLoadingEvent event){
        RegistryKey<Biome> biomeKey = RegistryKey.create(Registry.BIOME_REGISTRY, event.getName());
        if (BiomeDictionary.hasType(biomeKey, BiomeDictionary.Type.OVERWORLD)){
            event.getGeneration().addStructureStart(ModFeatures.towerAir);
            event.getGeneration().addStructureStart(ModFeatures.towerWater);
            event.getGeneration().addStructureStart(ModFeatures.towerEarth);
        } else if (BiomeDictionary.hasType(biomeKey, BiomeDictionary.Type.NETHER)){
            event.getGeneration().addStructureStart(ModFeatures.towerFire);
        } else if (BiomeDictionary.hasType(biomeKey, BiomeDictionary.Type.END)){
            event.getGeneration().addStructureStart(ModFeatures.towerOrder);
            event.getGeneration().addStructureStart(ModFeatures.towerChaos);
        }

    }

    @SubscribeEvent
    static void loadWorld(WorldEvent.Load event){
        if (event.getWorld().isClientSide())
            return;
        ServerWorld world = (ServerWorld) event.getWorld();
        RegistryKey<World> dim = world.dimension();
        if(world.getChunkSource().getGenerator() instanceof FlatChunkGenerator
            && dim.equals(World.OVERWORLD))
            return;
        world.getChunkSource().generator.getSettings().structureConfig().put(ModFeatures.tower, Tower.separation);
    }
}

File: src/main/kotlin/arcana/common/events/WorldEvents.kt
package arcana.common.events

import arcana.common.CommonData
import arcana.common.aspects.AspectList
import arcana.common.aspects.Aspects
import arcana.common.aspects.Aspects.EARTH
import arcana.common.aspects.Aspects.FIRE
import arcana.common.aspects.Aspects.WATER
import net.minecraft.util.RegistryKey
import net.minecraft.util.registry.Registry
import net.minecraft.world.biome.Biome.Category
import net.minecraftforge.common.BiomeDictionary
import net.minecraftforge.common.BiomeDictionary.Type.COLD
import net.minecraftforge.common.BiomeDictionary.Type.HOT
import net.minecraftforge.event.world.BiomeLoadingEvent
import net.minecraftforge.eventbus.api.SubscribeEvent
import net.minecraftforge.fml.common.Mod.EventBusSubscriber

@EventBusSubscriber
object WorldEvents {
    @SubscribeEvent
    fun loadBiome(event: BiomeLoadingEvent) {
        if (event.name == null) //it is possible, but when?
            return
        val biomeKey = RegistryKey.create(Registry.BIOME_REGISTRY, event.name!!)
        val types = BiomeDictionary.getTypes(biomeKey)
        val aspects = AspectList()
        for (type: BiomeDictionary.Type in types) {
            aspects.add(when(type) {
                HOT -> AspectList().add(FIRE, 1)
                COLD -> AspectList().add(WATER, 1)
                else -> AspectList()
            })
        }
        aspects.add(when(event.category) {
            Category.PLAINS -> AspectList().add(EARTH, 2)
            else -> AspectList()
        })
        val sum = aspects.sum().toFloat()
        if (sum == 0f)
            return
        val vis = FloatArray(6) {i -> aspects.getAmount(Aspects.PRIMAL[i]) * 1f/20f * 1f/sum * 6}
        CommonData.BIOME_VIS[biomeKey] = vis
    }
}